apify-platform/deploying section (#351)
apify/apify-docs: content/academy/apify_platform/deploying/inputs_outputs.md

---
title: Inputs & outputs
description: Learn to accept input into your actor, do something with it, then return output. Actors can be written in any language, so this concept is language agnostic.
menuWeight: 1
paths:
    - apify-platform/deploying/inputs-outputs
---

# [](#inputs-outputs) Inputs & outputs

Most of the time when you're creating a project, you are expecting some sort of input off of which your software will run off of. Often times as well, you want to provide some sort of output once your software has completed running. With Apify, it is extremely easy to take in inputs and deliver outputs.

An important thing to understand regarding inputs and outputs is that they are read/written differently depending on where the actor is running:

> - If your actor is running locally, the inputs/outputs are usually provided in the filesystem, and environment variables are injected either by you, the developer, and by the Apify CLI by running the project with the `apify run` command.

Reviewer: Just an idea. What about moving the generic and installation part of the Apify CLI (in deploying) to glossary/tools? That way, we could link to it, for example from here. And the deploying lesson would reference it for installation and only continue with the deploying part.

I think this could be useful, because we will be talking about Apify CLI in many places, so we could have one "crossroads" for the user. We could even have bi-directional links - e.g. link from deploying to Apify CLI concept, but also link from the concept to the deploying section.

Author: really good idea

- While running in a Docker container on the platform, environment variables are automatically injected, and inputs & outputs are provided and modified using Apify's REST API.

## [](#accepting-input) Accepting input

There are multiple ways to accept input into your project. The option you go with depends on the language you have written your project in. If you are using Node.js for your repo's code, you can use the [Apify SDK](https://www.npmjs.com/package/apify). Otherwise, you can use the useful environment variables automatically set up for you by Apify to write utility functions which read the actor's input and return it.

### Accepting input with the Apify SDK

Since we're using Node.js, let's install the **apify** package by running the following command:

```shell
npm install apify
```

Now, let's import `Apify` and use the `Apify.getInput()` function to grab our input.

```JavaScript
// index.js
import Apify from 'apify';

const input = await Apify.getInput();
console.log(input);
```

> If we run this right now, we'll see **null** in our terminal - this is because we never provided any sort of test input, which should be provided in the default key-value store. The `Apify.getInput()` function has detected that there is no **apify_storage** folder and generated one for us. We'll now add an **INPUT.json** file within **apify_storage/key_value_stores/default** to match what we're expecting in our code.

Reviewer: A screenshot or full path in a codeblock would be nice, just to make sure the users understand the path. They're often more confused than one would expect.

```JSON
{
    "numbers": [5, 5, 5, 5]
}
```

> Then we can modify our code to grab the input and use it to generate a solution which is logged to the console.

Reviewer: ```suggestion
Then we can add our example project code from earlier. It will grab the input and use it to generate a solution which is logged to the console.
```

```JavaScript
// index.js
import Apify from 'apify';

const { numbers } = await Apify.getInput();

> 

Reviewer: ```suggestion
// This is our example project code from earlier.
// We will use the Apify input as its input.
```

const addAllNumbers = (...nums) => nums.reduce((total, curr) => (total += curr));

> 

Reviewer: ```suggestion
```

This is only a code style thing, but we generally prefer visually grouping related functionality so I would suggest removing the blank line between adding the numbers as the whole thing is the example project.
<img width="802" alt="CleanShot 2022-04-28 at 11 33 47@2x" src="https://user-images.githubusercontent.com/23726914/165723629-e4462c1f-e181-4748-8a73-afead3d0e31e.png">

const solution = addAllNumbers(...numbers);

> 

Reviewer: ```suggestion
// ---------
```

console.log(solution);
```

Cool! When we run `node index.js`, we see **20**.

### Accepting input without the Apify SDK

Alternatively, when writing in a language other than JavaScript, we can create our own `get_input()` function which utilizes the Apify API. For this example, we are using the [Apify Client]({{@link apify_platform/getting_started/apify_client.md}}) for Python to access the API.

```Python
# index.py
from apify_client import ApifyClient
from os import environ
import json

client = ApifyClient(token='YOUR_TOKEN')

# If being run on the platform, the "APIFY_IS_AT_HOME" environment variable
# will be "1". Otherwise, it will be undefined/None
def is_on_apify ():
    if (environ.get('APIFY_IS_AT_HOME')):
        return True
    return False

# Get the input
def get_input ():
    if (not is_on_apify()):
        input = open('./apify_storage/key_value_stores/default/INPUT.json')
        return json.load(input)

    kv_store = client.key_value_store(environ.get('APIFY_DEFAULT_KEY_VALUE_STORE_ID'))
    kv_store.get_record('INPUT', format='json')

> 

Reviewer: ```suggestion
# This is our example Python project code from earlier.
# We will use the Apify input as its input.
```

def add_all_numbers (nums):
    total = 0

    for num in nums:
        total += num

    return total

solution = add_all_numbers(get_input()['numbers'])

> 

Reviewer: ```suggestion
# ------------
```

print(solution)
```

> For a better understanding of API endpoints reading and modifying key-value stores, check the [official API reference](https://docs.apify.com/api/v2#/reference/key-value-stores).

## [](#writing-output) Writing output

> Similarly to reading input, you can write the actor's output either by using the Apify SDK in Node.js, or by manually writing a utility function to do so.

Reviewer: I'm wondering if this would be a good opportunity to quickly explain datasets and key-value stores and link to their docs. Without that, the output saving is a bit magical and begs the question where does it actually save.

### Writing output with the Apify SDK

In the SDK, we can write to the key-value store with the `Apify.setValue()` function. Let's go ahead and write the solution of the `addAllNumbers()` function to the key-value store using it:

```JavaScript
// index.js
import Apify from 'apify';

const { numbers } = await Apify.getInput();

> 

Reviewer: ```suggestion
// This is our example project code from earlier.
// We will use the Apify input as its input.
```

const addAllNumbers = (...nums) => nums.reduce((total, curr) => (total += curr));

> 

Reviewer: ```suggestion
```

const solution = addAllNumbers(...numbers);

> 

Reviewer: ```suggestion
// And save its output to the key-value store.
```

> await Apify.setValue('OUTPUT', { solution });

Reviewer: I know that a KVS seems like the obvious option here, but could we do a Dataset instead? I'm asking because output schema (pretty output in app) is only working with datasets now.

Author: I agree it should be dataset, just used the KVStore because Lukas used KVStore in his Rust example which I translated to JS, then translated that to Python haha

```

### Writing output without the Apify SDK

Just as with the custom `get_input()` utility function, you can write a custom `set_output()` function as well if you cannot use the Apify SDK.

> You can read and write your output anywhere; however, it is standard practice to use a folder named **apify_storage**.

> ```Python

Reviewer: Please update this in line with the JS updates as well.

- While running in a Docker container on the platform, environment variables are automatically injected, and inputs & outputs are provided and modified using Apify's REST API.

## [](#about-storage) A bit about storage

> There are two different places you can read/write your inputs/outputs: to the [key-value store](https://docs.apify.com/storage/key-value-store), or to the [dataset](https://docs.apify.com/storage/dataset). They key-value store can be used to store any sort of unorganized/unrelated data in any formats, while the data pushed to a dataset generally has the same fields/is of the same (or similar) schema. Each actor's run is allocated both a default dataset and key-value store.

Reviewer: ```suggestion
There are two different places you can read/write your inputs/outputs: to the [key-value store](https://docs.apify.com/storage/key-value-store), or to the [dataset](https://docs.apify.com/storage/dataset). They key-value store can be used to store any sort of unorganized/unrelated data in any formats, while the data pushed to a dataset typically resembles a table with columns (fields) and rows (items). Each actor's run is allocated both a default dataset and key-value store.
```
